### PR TITLE
Add map and filter_map to Ref and RefMut

### PR DESCRIPTION
**Pull request by Aursen**
Add utility functions based on [solana-nostd-entrypoint](https://github.com/cavemanloverboy/solana-nostd-entrypoint/commit/2f9d1173e84f4106fc4ff4282115ac26364f4670)

### CODE REVIEW

**Comment by Aursen** [comment]
Would you like me to add some tests?

**Comment by febo** [comment]
> Would you like me to add some tests?

That would be great, thanks!

**Comment by Aursen** [comment]
Done

File: sdk/pinocchio/src/account_info.rs [hunk 1]
```diff
@@ -373,17 +382,66 @@ const DATA_SHIFT: u8 = 0;
 
 /// Reference to account data or lamports with checked borrow rules.
 pub struct Ref<'a, T: ?Sized> {
-    value: &'a T,
+    value: NonNull<T>,
     state: NonNull<u8>,
     /// Indicates the type of borrow (lamports or data) by representing the
     /// shift amount.
     borrow_shift: u8,
+    marker: PhantomData<&'a T>,
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:390]: Could we add a comment to explain the need for the `PhantonData` – e.g.,:
```rust
/// The `value` raw pointer is only valid while the `&'a T` lives so we claim
/// to hold a reference to it.
marker: PhantomData<&'a T>,
```

File: sdk/pinocchio/src/account_info.rs [hunk 2]
```diff
@@ -373,17 +382,66 @@ const DATA_SHIFT: u8 = 0;
...
+        // stop the decrement via Drop
+        core::mem::forget(orig);
+
+        Ref {
+            state,
+            value,
+            borrow_shift,
+            marker: PhantomData,
+        }
+    }
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:413]: I wonder if we should use a `ManuallyDrop` here – it seems to suite better when transferring memory ownership according to the docs.

```suggestion
    #[inline]
    pub fn map<U: ?Sized, F>(orig: Ref<'a, T>, f: F) -> Ref<'a, U>
    where
        F: FnOnce(&T) -> &U,
    {
        // Avoid decrementing the borrow flag on Drop.
        let orig = ManuallyDrop::new(orig);

        Ref {
            value: NonNull::from(f(&*orig)),
            state: orig.state,
            borrow_shift: orig.borrow_shift,
            marker: PhantomData,
        }
    }
```

File: sdk/pinocchio/src/account_info.rs [hunk 3]
```diff
@@ -373,17 +382,66 @@ const DATA_SHIFT: u8 = 0;
...
+                Ok(Ref {
+                    value,
+                    state,
+                    borrow_shift,
+                    marker: PhantomData,
+                })
+            }
+            None => Err(orig),
+        }
+    }
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:438]: Same here with a `ManuallyDrop`.

```suggestion
    #[inline]
    pub fn filter_map<U: ?Sized, F>(orig: Ref<'a, T>, f: F) -> Result<Ref<'a, U>, Self>
    where
        F: FnOnce(&T) -> Option<&U>,
    {
        // Avoid decrementing the borrow flag on Drop.
        let orig = ManuallyDrop::new(orig);

        match f(&*orig) {
            Some(value) => Ok(Ref {
                value: NonNull::from(value),
                state: orig.state,
                borrow_shift: orig.borrow_shift,
                marker: PhantomData,
            }),
            None => Err(ManuallyDrop::into_inner(orig)),
        }
    }
```

File: sdk/pinocchio/src/account_info.rs [hunk 4]
```diff
@@ -402,22 +460,71 @@ const DATA_MASK: u8 = 0b_1111_0111;
 
 /// Mutable reference to account data or lamports with checked borrow rules.
 pub struct RefMut<'a, T: ?Sized> {
-    value: &'a mut T,
+    value: NonNull<T>,
     state: NonNull<u8>,
     /// Indicates the type of borrow (lamports or data) by representing the
     /// mutable borrow mask.
     borrow_mask: u8,
+    marker: PhantomData<&'a mut T>,
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:468]: We could add a comment here as well.

```rust
/// The `value` raw pointer is only valid while the `&'a mut T` lives so we claim
/// to hold a reference to it.
marker: PhantomData<&'a mut T>,
```

File: sdk/pinocchio/src/account_info.rs [hunk 5]
```diff
@@ -402,22 +460,71 @@ const DATA_MASK: u8 = 0b_1111_0111;
...
+        // stop the decrement via Drop
+        core::mem::forget(orig);
+
+        RefMut {
+            value,
+            state,
+            borrow_mask,
+            marker: PhantomData,
+        }
+    }
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:491]: Same here about the `ManuallyDrop`.

```suggestion
    #[inline]
    pub fn map<U: ?Sized, F>(orig: RefMut<'a, T>, f: F) -> RefMut<'a, U>
    where
        F: FnOnce(&mut T) -> &mut U,
    {
        // Avoid decrementing the mutable borrow flag on Drop.
        let mut orig = ManuallyDrop::new(orig);

        RefMut {
            value: NonNull::from(f(&mut *orig)),
            state: orig.state,
            borrow_mask: orig.borrow_mask,
            marker: PhantomData,
        }
    }
```

File: sdk/pinocchio/src/account_info.rs [hunk 6]
```diff
@@ -402,22 +460,71 @@ const DATA_MASK: u8 = 0b_1111_0111;
...
+                Ok(RefMut {
+                    value,
+                    state,
+                    borrow_mask,
+                    marker: PhantomData,
+                })
+            }
+            None => Err(orig),
+        }
+    }
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:516]: And here as well.

```suggestion
    #[inline]
    pub fn filter_map<U: ?Sized, F>(orig: RefMut<'a, T>, f: F) -> Result<RefMut<'a, U>, Self>
    where
        F: FnOnce(&mut T) -> Option<&mut U>,
    {
        // Avoid decrementing the mutable borrow flag on Drop.
        let mut orig = ManuallyDrop::new(orig);

        match f(&mut *orig) {
            Some(value) => {
                let value = NonNull::from(value);

                Ok(RefMut {
                    value,
                    state: orig.state,
                    borrow_mask: orig.borrow_mask,
                    marker: PhantomData,
                })
            }
            None => Err(ManuallyDrop::into_inner(orig)),
        }
    }
```

File: sdk/pinocchio/src/account_info.rs [hunk 7]
```diff
@@ -427,3 +534,129 @@ impl<'a, T: ?Sized> Drop for RefMut<'a, T> {
...
+
+        drop(new_ref);
+
+        assert_eq!(state, 0 << LAMPORTS_SHIFT);
+    }
+
+    #[test]
+    fn test_data_ref_mut() {
+        let data: [u8; 4] = [0, 1, 2, 3];
+        let state = 0 | 0b_0000_1000;
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:613]: This could be just `let state = 0b_0000_1000;`

File: sdk/pinocchio/src/account_info.rs [hunk 8]
```diff
@@ -427,3 +534,129 @@ impl<'a, T: ?Sized> Drop for RefMut<'a, T> {
...
+        drop(new_ref);
+
+        assert_eq!(data, [4, 1, 2, 3]);
+        assert_eq!(state, 0);
+    }
+
+    #[test]
+    fn test_lamports_ref_mut() {
+        let lamports: u64 = 10000;
+        let state = 0 | 0b_1000_0000;
```
Comment (febo) [inline sdk/pinocchio/src/account_info.rs:640]: Here as well, just `let state = 0b_0000_1000;`

**Review by febo — APPROVED** [review]

**Pull request merged**